plugins/permission-node/src/integration/createPermissionIntegrationRouter.ts

```diff
@@ -92,84 +119,103 @@ const applyConditions = <TResource>(
- * captured here as `resourceType`, a string identifier (`catalog-entity` in this example) that can
- * be provided with permission definitions. This is merely a _type_ to verify that conditions in an
- * authorization policy are constructed correctly, not a reference to a specific resource.
+ * To make this concrete, we can use the Backstage software catalog as an
+ * example. The catalog has conditional rules around access to specific
+ * _entities_ in the catalog. The _type_ of resource is captured here as
+ * `resourceType`, a string identifier (`catalog-entity` in this example) that
+ * can be provided with permission definitions. This is merely a _type_ to
+ * verify that conditions in an authorization policy are constructed correctly,
+ * not a reference to a specific resource.
  *
- * The `rules` parameter is an array of {@link PermissionRule}s that introduce conditional
- * filtering logic for resources; for the catalog, these are things like `isEntityOwner` or
- * `hasAnnotation`. Rules describe how to filter a list of resources, and the `conditions` returned
- * allow these rules to be applied with specific parameters (such as 'group:default/team-a', or
+ * The `rules` parameter is an array of {@link PermissionRule}s that introduce
+ * conditional filtering logic for resources; for the catalog, these are things
+ * like `isEntityOwner` or `hasAnnotation`. Rules describe how to filter a list
+ * of resources, and the `conditions` returned allow these rules to be applied
+ * with specific parameters (such as 'group:default/team-a', or
  * 'backstage.io/edit-url').
  *
- * The `getResource` argument should load a resource by reference. For the catalog, this is an
- * {@link @backstage/catalog-model#EntityRef}. For other plugins, this can be any serialized format.
- * This is used to construct the `createPermissionIntegrationRouter`, a function to add an
- * authorization route to your backend plugin. This route will be called by the `permission-backend`
- * when authorization conditions relating to this plugin need to be evaluated.
+ * The `getResources` argument should load resources based on a reference
+ * identifier. For the catalog, this is an
+ * {@link @backstage/catalog-model#EntityRef}. For other plugins, this can be
+ * any serialized format. This is used to construct the
+ * `createPermissionIntegrationRouter`, a function to add an authorization route
+ * to your backend plugin. This function will be called by the
+ * `permission-backend` when authorization conditions relating to this plugin
+ * need to be evaluated.
  *
  * @public
  */
 export const createPermissionIntegrationRouter = <TResource>(options: {
   resourceType: string;
   rules: PermissionRule<TResource, any>[];
-  getResource: (resourceRef: string) => Promise<TResource | undefined>;
-}): Router => {
-  const { resourceType, rules, getResource } = options;
+  getResources: (
```

Comment:
I've considered a couple of options for the new return value here. An array is less explicit, but much easier for consumers to build correctly. Conversely, a record object mapping the input resource refs to their corresponding resources is much more explicit, but more work to put together. It also has its own pitfalls - in particular, it means that implementers must carefully handle normalized resource refs, which is fiddly, error prone, and easy to forget.

This PR currently proposes an array return value, but a prior commit shows the record return value. There's also examples of implementing both the array and record approaches for the catalog-backend: array, record.

Reply:
i think i like the array more, reminds me of graphql dataloader